### PR TITLE
Agent Pip Config

### LINKED ISSUE #6517: pip config: agent side
Implement agent side of the [pip config design](https://github.com/inmanta/designs/blob/master/modules/pip_config.md). Do not include changelog entries but if new breaking changes come to light, communicate them at #6518.

### PR DESCRIPTION
**Pull request by wouterdb**
# Description

added agent config

closes  #6517 
# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [ ] Code is clear and sufficiently documented
- [ ] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

File: src/inmanta/data/__init__.py [hunk 1]
```diff
@@ -2000,7 +2007,7 @@ def _get_value(cls, value: object) -> object:
         if isinstance(value, dict):
             return json_encode(value)
 
-        if isinstance(value, DataDocument) or issubclass(value.__class__, DataDocument):
+        if isinstance(value, DataDocument) or issubclass(value.__class__, DataDocument) or isinstance(value, BaseModel):
```
Comment (wouterdb) [inline src/inmanta/data/__init__.py:2010]: !!! Added the ability to use pydantic objects towards the DB

File: src/inmanta/data/__init__.py [hunk 2]
```diff
@@ -5208,7 +5217,10 @@ async def create_for_partial_compile(
         the partial compile, i.e. not present in rids_in_partial_compile.
         """
         query = f"""
-            WITH base_version_exists AS (
+            WITH base_version_record AS (
```
Comment (wouterdb) [inline src/inmanta/data/__init__.py:5220]: @arnaudsjs I'm not very confident I did this in a good way, I just jiggled it until it passes the test

File: src/inmanta/data/model.py [hunk 3]
```diff
@@ -741,3 +741,56 @@ def to_dao(self, env: uuid) -> "data.DiscoveredResource":
...
+
+
+def hyphenize(field: str) -> str:
+    """Alias generator to convert python names (with `_`) to config file name (with `-`)"""
+    return field.replace("_", "-")
+
+
+@stable_api
+# This is part of both the config file schema and the api schema
+class PipConfig(BaseModel):
```
Comment (wouterdb) [inline src/inmanta/data/model.py:753]: I had to move this to avoid import loops

File: src/inmanta/export.py [hunk 4]
```diff
@@ -546,27 +548,40 @@ def commit_resources(
...
+            return result
+
+        # Backward compatibility with ISO6 servers
+        result = do_put(pip_config=pip_config)
+        if (
+            result.code == 400
+            and isinstance(result.result, dict)
+            and "Invalid request: request contains fields {'pip_config'}" in result.result.get("message", "")
+        ):
+            result = do_put()
```
Comment (wouterdb) [inline src/inmanta/export.py:584]: I will produce a warning here

File: tests/data/modules/test_resources/model/_init.cf [hunk 5]
```diff
@@ -0,0 +1,7 @@
+entity Resource extends std::PurgeableResource:
```
Comment (wouterdb) [inline tests/data/modules/test_resources/model/_init.cf:1]: I moved this out of a testcase to make testing faster

File: src/inmanta/agent/agent.py [hunk 6]
```diff
@@ -1385,10 +1388,20 @@ async def _install(self, sources: list[ModuleSource], requirements: Sequence[str
...
                 list(pkg_resources.parse_requirements(requirements)),
-                PipConfig(use_system_config=True),
+                pip_config,
             )
             await loop.run_in_executor(self.thread_pool, self._loader.deploy_version, sources)
 
+    async def _get_pip_config(self, environment: uuid.UUID, version: int) -> PipConfig:
+        response = await self._client.get_pip_config(tid=environment, version=version)
+        if response.code != 200:
+            raise Exception("Could not get pip config from server " + str(response.result))
```
Comment (wouterdb) [inline src/inmanta/agent/agent.py:1399]: I'm not 100% clear what I should report here.

File: src/inmanta/protocol/methods_v2.py [hunk 7]
```diff
@@ -1285,6 +1284,23 @@ def get_source_code(tid: uuid.UUID, version: int, resource_type: str) -> List[mo
...
 
+@typedmethod(
+    path="/pip/config/<version>",
+    operation="GET",
+    agent_server=True,
+    arg_options=methods.ENV_OPTS,
+    client_types=[ClientType.agent],
+    api_version=2,
+)
+def get_pip_config(tid: uuid.UUID, version: int) -> model.PipConfig:
```
Comment (wouterdb) [inline src/inmanta/protocol/methods_v2.py:1296]: Should this return `Optional[PipConfig]` ? If I do this, the agent knows when the model is compiled with an too old version, not it just gets the fallback config?

File: src/inmanta/data/model.py [hunk 8]
```diff
@@ -741,3 +741,74 @@ def to_dao(self, env: uuid) -> "data.DiscoveredResource":
...
+    use_system_config: bool = False
+
+    def has_source(self) -> bool:
+        """Can this config get packages from anywhere?"""
+        return bool(self.index_url) or self.use_system_config
+
+    def assert_has_source(self, reason: str) -> None:
+        """Ensure this index has a valid package source, otherwise raise exception"""
+        # Import loop
+        import inmanta.env
```
Comment (wouterdb) [inline src/inmanta/data/model.py:801]: any suggestions how to clean this up?

File: src/inmanta/agent/agent.py [hunk 9]
```diff
@@ -1385,10 +1388,20 @@ async def _install(self, sources: list[ModuleSource], requirements: Sequence[str
...
+        pip_config = response.result["data"]
+        if pip_config is None:
```
Comment (wouterdb) [inline src/inmanta/agent/agent.py:1401]: This can not currently happen, see below

File: src/inmanta/data/__init__.py (hunk 2 above)
Comment (arnaudsjs) [reply src/inmanta/data/__init__.py:5220]: Looks good to me.

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/protocol/methods_v2.py (hunk 7 above)
Comment (wouterdb) [reply src/inmanta/protocol/methods_v2.py:1296]: changed

File: src/inmanta/server/services/orchestrationservice.py [hunk 10]
```diff
@@ -901,6 +931,7 @@ async def put_partial(
         version_info: Optional[JsonType] = None,
         resource_sets: Optional[Dict[ResourceIdStr, Optional[str]]] = None,
         removed_resource_sets: Optional[List[str]] = None,
+        pip_config: Optional[PipConfig] = None,
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/server/services/orchestrationservice.py:933]: added

File: src/inmanta/protocol/methods.py [hunk 11]
```diff
@@ -511,11 +512,12 @@ def put_version(
     tid: uuid.UUID,
     version: int,
     resources: list,
-    resource_state: dict = {},
-    unknowns: Optional[list] = None,
+    resource_state: Dict[model.ResourceIdStr, Literal[ResourceState.available, ResourceState.undefined]] = {},
```
Comment (wouterdb) [inline src/inmanta/protocol/methods.py:515]: This is perhaps a bit risky?

**Review by arnaudsjs — APPROVED** [review]

File: tests/agent_server/test_agent_code_loading.py [hunk 12]
```diff
@@ -218,18 +245,28 @@ def test():
...
+    assert res.code == 200
+
+    res = await client.upload_code_batched(tid=environment, id=version, resources={"test::Test": sources})
     assert res.code == 200
 
     agent: Agent = await agent_factory(
         environment=environment, agent_map={"agent1": "localhost"}, hostname="host", agent_names=["agent1"], code_loader=True
     )
 
     # Set the PIP_INDEX_URL, because the agent doesn't have support for custom indexes yet
```
Comment (sanderr) [inline tests/agent_server/test_agent_code_loading.py:266]:
```suggestion
```

Handy that there was already a test prepared for this.

**Review by sanderr — CHANGES_REQUESTED** [review]
Didn't you say you wanted to return a warning for `put_version` and `put_partial` if no pip config is set? (The opposite of the warning you already have.)

File: src/inmanta/data/model.py (hunk 8 above)
Comment (wouterdb) [reply src/inmanta/data/model.py:788]: done

File: src/inmanta/data/model.py [hunk 13]
```diff
@@ -16,11 +16,12 @@
     Contact: code@inmanta.com
 """
 import datetime
+import typing
 import uuid
 from collections import abc
 from enum import Enum
 from itertools import chain
-from typing import ClassVar, Dict, List, NewType, Optional, Self, Union
+from typing import ClassVar, Dict, List, NewType, Optional, Self, Sequence, Union
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/data/model.py:24]: it is, right?

File: src/inmanta/server/services/orchestrationservice.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/server/services/orchestrationservice.py:928]: https://github.com/inmanta/designs/pull/131

File: src/inmanta/data/model.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/data/model.py:24]: Sorry, typo. Should be used from `collections.abc`.

File: src/inmanta/agent/agent.py [hunk 14]
```diff
@@ -1385,10 +1389,17 @@ async def _install(self, sources: list[ModuleSource], requirements: Sequence[str
...
+                pip_config,
             )
             await loop.run_in_executor(self.thread_pool, self._loader.deploy_version, sources)
 
+    async def _get_pip_config(self, environment: uuid.UUID, version: int) -> PipConfig:
+        response = await self._client.get_pip_config(tid=environment, version=version)
+        if response.code != 200:
+            raise Exception("Could not get pip config from server " + str(response.result))
+        pip_config = response.result["data"]
+        return PipConfig(**pip_config)
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/agent/agent.py:1401]: yes, thx. I think I had it as some point, then removed it, very good catch.

**Comment by wouterdb** [comment]
> Didn't you say you wanted to return a warning for `put_version` and `put_partial` if no pip config is set? (The opposite of the warning you already have.)

I had wanted to, but the older compilers would not print it anyways. 
(Warnings support there is very limited)

**Review by sanderr — APPROVED** [review]
Just a reminder that Bart asked to hold off merging potentially breaking changes until the dust has settled.

**Pull request closed without merging**